elytron/src/main/java/org/wildfly/extension/elytron/JdbcRealmDefinition.java

```diff
@@ -653,33 +688,46 @@ private AttributeMapper[] resolveAttributeMappers(OperationContext context, Mode
         }
     }
 
-    private static KeyMapper resolveKeyMappers(OperationContext context, ModelNode authenticationQueryNode) throws OperationFailedException {
-        KeyMapper keyMapper = null;
-
-        for (String name : authenticationQueryNode.keys()) {
-            ModelNode propertyNode = authenticationQueryNode.require(name);
+    private static class JdbcRealmDefinitionValidation implements OperationStepHandler {
 
-            if (!propertyNode.isDefined()) {
-                continue;
+        @Override
+        public void execute(OperationContext context, ModelNode operation) throws OperationFailedException {
+            List<ModelNode> principalQueries = PrincipalQueryAttributes.PRINCIPAL_QUERIES_7_0.resolveModelAttribute(context, operation).asList();
+            for (ModelNode modelNode : principalQueries) {
+                checkKeyMappers(context, modelNode);
             }
+        }
 
-            PasswordMapperObjectDefinition mapperResource = PrincipalQueryAttributes.SUPPORTED_PASSWORD_MAPPERS.get(name);
+        private static void checkKeyMappers(OperationContext context, ModelNode authenticationQueryNode) throws OperationFailedException {
 
-            if (mapperResource == null) {
-                continue;
-            }
+            KeyMapper keyMapper = null;
 
-            if (keyMapper != null) {
-                throw ROOT_LOGGER.jdbcRealmOnlySingleKeyMapperAllowed();
-            }
+            for (String name : authenticationQueryNode.keys()) {
+                ModelNode propertyNode = authenticationQueryNode.require(name);
+
+                if (!propertyNode.isDefined()) {
+                    continue;
+                }
 
-            try {
-                keyMapper = mapperResource.toPasswordKeyMapper(context, propertyNode);
-            } catch (InvalidKeyException e) {
-                throw new OperationFailedException("Invalid key type.", e);
+                PasswordMapperObjectDefinition mapperResource = PrincipalQueryAttributes.SUPPORTED_PASSWORD_MAPPERS.get(name);
+
+                if (mapperResource == null) {
+                    continue;
+                }
+
+                if (keyMapper != null) {
+                    throw ROOT_LOGGER.jdbcRealmOnlySingleKeyMapperAllowed();
+                }
+
+                try {
+                    keyMapper = mapperResource.toPasswordKeyMapper(context, propertyNode);
```

Comment:
This is not allowed. The toPasswordKeyMapper implementations do work that is not meant for Stage.MODEL (all the resolveModelAttribute calls.

Reply:
That said, my impression reading this is that it's not necessary to call this in Stage.MODEL. This entire try/catch can go away. Looking (quickly) at all the impls of this interface, none of them do anything that is relevant to Stage.MODEL.

There's separate logic called in populateModel that validates the individual mapper configs, so the various things done in the toPasswordKeyMapper impls is not a necessary part of the Stage.MODEL validation. All that's needed here is checking that there is only config element.

Reply:
Hello Brian! Thank you for your revision. You mean, that I need do this validation without using the `toPassword...` method?